node/libs/protobuf/src/build.rs

```diff
@@ -81,5 +82,36 @@ macro_rules! impl_reflect_message {
     };
 }
 
+/// Implements `serde::Serialize` compatible with protobuf json encoding.
+#[macro_export]
+macro_rules! impl_serde_serialize {
```

Comment:
Just to understand better: Do we *want* for all generated messages to implement `Serialize` / `Deserialize`, or is it just the necessity because they are embedded into server types? If it's the latter, we could try [something like](https://serde.rs/field-attrs.html#with) `#[serde(with = "zksync_protobuf::serde")]` (i.e., provide a module that implements serialization / deserialization logic; essentially, neatly packaged `serde_serialize` / `serde_deserialize`). In this case, generated messages won't implement `Serialize` / `Deserialize`, but it would be possible to embed them into server types specifying this attribute.

Reply:
I don't really care, but I think it would be more straightforward to force users to use the well defined serialization rather than giving them opportunity to consider having a custom one. In fact the user is expected to define their own struct which maps to the generated struct (by implementing ProtoFmt) and embed THAT struct, which will need to have custom serde::Serialize/Deserialize implementation anyway and in this case there is no actual need for the generated struct to implement serde traits directly at all.

Reply:
Sorry, I've realized after making the comment that you're defining `serde` impls for the generated types, but we'd want to embed not them, but rather their wrappers; that's certainly a source of confusion. But still, 

> which will need to have custom serde::Serialize/Deserialize implementation anyway

Not necessarily; it can be embedded into other types with `#[serde(with)]` using the approach I've mentioned. It'd look something like:

```rust
// mod `zksync_protobuf::serde`

use crate::ProtoFmt;
use prost_reflect::{prost::Message, ReflectMessage};
use serde::de::Error as _;

/// Implementation of serde::Serialize for arbitrary ReflectMessage.
pub fn serialize<T: ProtoFmt, S: serde::Serializer>(
    x: &T,
    s: S,
) -> Result<S::Ok, S::Error> {
    let opts = prost_reflect::SerializeOptions::new();
    x.build().transcode_to_dynamic().serialize_with_options(s, &opts)
}


/// Implementation of serde::Deserialize for arbitrary ReflectMessage.
pub fn deserialize<'de, T: ProtoFmt, D: serde::Deserializer<'de>>(
    d: D,
) -> Result<T, D::Error> {
    let mut p = <T::Proto>::default();
    let msg = prost_reflect::DynamicMessage::deserialize(p.descriptor(), d)?;
    p.merge(msg.encode_to_vec().as_slice()).unwrap();
    let p = T::read(&p).map_err(<D::Error>::custom)?;
    Ok(p)
}
```

Example of usage (on the example of `roles` types):

```rust
#[derive(Debug, Deserialize, Serialize)]
struct Test {
    #[serde(with = "zksync_protobuf::serde")]
    key: PublicKey,
    #[serde(with = "zksync_protobuf::serde")]
    signature: AggregateSignature,
}
```

Similarly, you can reinterpret `impl_serde_*` macros, so that they derive the corresponding traits for the type implementing `ProtoFmt` directly, something like

```rust
struct MyType { /* ... */ }

impl ProtoFmt for MyType { /* ... */ }

zksync_protobuf::serde::derive_based_on_proto_fmt!(MyType);
// ^ Derives `Serialize` and `Deserialize` using `serialize` and `deserialize` functions
```

Reply:
done, I've created a serde module with serialize and deserialize methods.